sw/device/lib/testing/usb_logging.c

```diff
@@ -0,0 +1,549 @@
+  /**
+   * Number of used bytes; index at which next write shall occur.
+   */
+  uint16_t bytes_used;
+  /**
+   * Data buffer holding logging data.
+   */
+  uint8_t data[kUSBLogBufferSize];
+};
+
+// Forwards reference to logging context.
+typedef struct usb_logging_ctx usb_logging_ctx_t;
+
+// Per-stream state information.
+typedef struct usb_logging_stream {
+  /**
+   * Context pointer; for use within callback handler.
+   */
+  usb_logging_ctx_t *ctx;
+  /**
+   * Reliable delivery required?
+   */
+  bool reliable;
+  /**
+   * Remapping for non-printable characters required?
+   */
+  bool remap;
+  /**
+   * The USB device endpoint that we're using.
+   */
+  uint8_t ep;
+  /**
+   * Currently sending a buffer?
+   */
+  bool sending;
+  /**
+   * The oldest logging buffer filled by this stream; this is either being
+   * sent (`sending` == true) or shall be the next to send. May be NULL iff
+   * this stream has no completed buffers to send.
+   */
+  usb_logging_buf_t *send_buf;
+  /**
+   * Time at which the most recently-populated buffer should be sent to the
+   * host; prevent log data being retained indefinitely in the event of no
+   * further log output.
+   */
+  ibex_timeout_t flush_time;
+  /**
+   * The current logging buffer; may be partially-filled or full (see
+   * `bytes_used` field), or NULL iff this stream has no pending logging data.
+   */
+  usb_logging_buf_t *wr_buf;
+} usb_logging_stream_t;
+
+// USB Logging context.
+struct usb_logging_ctx {
+  /**
+   * Access to usb_testutils layer.
+   */
+  usb_testutils_ctx_t *usbutils;
+  /**
+   * Number of logging streams.
+   */
+  uint8_t nstreams;
+  /**
+   * Linked-list of free buffers.
+   */
+  usb_logging_buf_t *free;
+  /**
+   * Per-stream state information.
+   */
+  usb_logging_stream_t streams[kUSBLogMaxStreams];
+  /**
+   * Pool of logging buffers; shared among all active streams.
+   */
+  usb_logging_buf_t buf[kUSBLogNumBuffers];
+};
+
+/**
+ * USB device context types.
+ */
+static usb_testutils_ctx_t usbdev;
+static usb_testutils_controlep_ctx_t usbdev_control;
+
+// Note: for now there is a single logging context.
+static usb_logging_ctx_t the_ctx;
+
+/**
+ * Pinmux handle
+ */
+static dif_pinmux_t pinmux;
+
+/**
+ * Internal development logging; be careful about using this if stdout has been
+ * redirected!
+ */
+static const bool kDevLogging = false;
+
+/**
+ * Character map, used to ensure that non-printable characters are dropped;
+ * isprint() available so this assumes ASCII and drops the control characters
+ * that could could issues with inappropriate terminal settings at the host.
+ */
+static const uint32_t kCtrlChars =
+    ((uint32_t)1u << '\t') | ((uint32_t)1u << '\n') | ((uint32_t)1u << '\r');
+static const uint32_t kMap[0x100u / 0x20u] = {
+    // suppress control characters and ASCII 127 (Delete)
+    kCtrlChars, ~0u, ~0u, 0x7fffffffu,
+    // top-bit set characters are harmless; propagate them unchanged.
+    ~0u, ~0u, ~0u, ~0u};
```

Comment:
Thinking some more about this, why do you think we should do this for USB if we don't do it for UART?

Reply:
It's there primarily because the USB logging functionality presently doesn't have an 'application' beyond serving as a top-level/integration test...if it ultimately serves just as console output as per UART and the other end of the connection is established and data collected within opentitanlib/tool then it won't be required.

Presently, though, if the logging data is gathered just using '/dev/ttyUSBn' on a Linux host with eg. 'cat' then any control characters that are sent can cause problems since the default tty settings respond to XON/XOFF (sw flow control), BRK and echo characters etc.